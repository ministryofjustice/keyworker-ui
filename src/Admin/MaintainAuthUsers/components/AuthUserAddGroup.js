import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import Select from '@govuk-react/select'
import Button from '@govuk-react/button'
import GridRow from '@govuk-react/grid-row'
import GridCol from '@govuk-react/grid-col'
import { BLACK, GREY_3 } from 'govuk-colours'
import { SPACING } from '@govuk-react/constants'
import { authGroupListType, errorType } from '../../../types'
import { lookupMeta } from '../../../govuk-helpers'

export const ButtonContainer = styled('div')`
  display: flex;
  flex-direction: row;
  button {
    margin-right: ${SPACING.SCALE_2};
  }
`

const AuthUserAddGroup = ({ groupFilterList, handleGroupAddChange, groupList, handleAdd, handleCancel, error }) => {
  const groupListWithoutCurrentGroups = groupFilterList.filter(
    filteredGroup => !groupList.some(currentGroup => currentGroup.groupCode === filteredGroup.groupCode)
  )

  const groupListOptions = groupListWithoutCurrentGroups
    ? groupListWithoutCurrentGroups.map(group => (
        <option key={`group_option_${group.groupCode}`} data-qa={`${group.groupCode}_option`} value={group.groupCode}>
          {group.groupName}
        </option>
      ))
    : []

  const groupsAvailable = groupListOptions && groupListOptions.length > 0

  return (
    <GridRow>
      <GridCol setWidth="two-thirds">
        <form onSubmit={handleAdd}>
          {!groupsAvailable && <div data-qa="no-groups">No groups available</div>}
          {groupsAvailable && (
            <Select
              label="Choose new group"
              mb={6}
              meta={lookupMeta('group', error)}
              htmlFor="group"
              input={{ id: 'group', name: 'group', onChange: handleGroupAddChange }}
            >
              <option key="choose" value="--">
                -- Select --
              </option>
              {groupListOptions}
            </Select>
          )}
          <ButtonContainer>
            {groupsAvailable && (
              <Button type="submit" data-qa="add-button">
                Add group
              </Button>
            )}
            <Button
              type="button"
              data-qa="cancel-button"
              onClick={handleCancel}
              buttonColour={GREY_3}
              buttonTextColour={BLACK}
            >
              Cancel
            </Button>
          </ButtonContainer>
        </form>
      </GridCol>
    </GridRow>
  )
}

AuthUserAddGroup.propTypes = {
  handleGroupAddChange: PropTypes.func.isRequired,
  handleAdd: PropTypes.func.isRequired,
  handleCancel: PropTypes.func.isRequired,
  groupFilterList: authGroupListType.isRequired,
  groupList: authGroupListType.isRequired,
  error: errorType.isRequired,
}

export default AuthUserAddGroup
