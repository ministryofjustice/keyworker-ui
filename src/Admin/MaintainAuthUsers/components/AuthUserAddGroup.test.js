import React from 'react'
import renderer from 'react-test-renderer'
import Label from '@govuk-react/label'
import AuthUserAddGroup from './AuthUserAddGroup'

const groups = [
  { groupCode: 'ROLE_1', groupName: 'Group 1' },
  { groupCode: 'ROLE_2', groupName: 'Group 2' },
  { groupCode: 'ROLE_3', groupName: 'Group 3' },
]

const stubFunc = () => {}

describe('Auth user add groups', () => {
  describe('no groups', () => {
    const wrapper = renderer.create(
      <AuthUserAddGroup
        groupList={groups}
        groupFilterList={[]}
        handleGroupAddChange={stubFunc}
        handleAdd={stubFunc}
        handleCancel={stubFunc}
        error=""
      />
    )
    it('should render correctly with user and no groups', () => {
      expect(wrapper.toJSON()).toMatchSnapshot()
    })

    it('should display message about no groups to add', () => {
      const noRowsMessage = wrapper.root.findByProps({ 'data-qa': 'no-groups' })
      // last table row should have message
      expect(noRowsMessage.props.children).toEqual('No groups available')
    })
  })

  it('should render the add group page with a filtered list of groups', () => {
    const component = renderer.create(
      <AuthUserAddGroup
        groupList={[{ groupCode: 'GROUP_2', groupName: 'Group 2' }]}
        groupFilterList={groups}
        handleAdd={jest.fn()}
        handleCancel={jest.fn()}
        handleGroupAddChange={jest.fn()}
        error=""
      />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })

  describe('errors', () => {
    const component = renderer.create(
      <AuthUserAddGroup
        groupList={[{ groupCode: 'GROUP_2', groupName: 'Group 2' }]}
        groupFilterList={groups}
        handleAdd={jest.fn()}
        handleCancel={jest.fn()}
        handleGroupAddChange={jest.fn()}
        error={[{ targetName: 'group', text: 'something went wrong' }]}
      />
    )
    it('should render an error on the select if error occurred', () => {
      expect(component.toJSON()).toMatchSnapshot()
    })
    it('should display error', () => {
      expect(component.root.findByType(Label).props.error).toEqual('something went wrong')
    })
  })
})
