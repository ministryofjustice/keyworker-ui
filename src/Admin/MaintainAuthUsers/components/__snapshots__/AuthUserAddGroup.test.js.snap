// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth user add groups errors should render an error on the select if error occurred 1`] = `
<div
  className="sc-Axmtr hjHhIB"
>
  <div
    className="sc-AxmLO czuJjP"
  >
    <form
      onSubmit={[MockFunction]}
    >
      <label
        className="sc-AxjAm iLFoNS"
        htmlFor="group"
      >
        <span
          className="sc-AxirZ lbZqdG"
        >
          Choose new group
        </span>
        <span
          className="sc-AxiKw MIKJW"
        >
          something went wrong
        </span>
        <select
          className="sc-AxhUy hIXZuA"
          id="group"
          name="group"
          onChange={[MockFunction]}
        >
          <option
            value="--"
          >
            -- Select --
          </option>
          <option
            data-qa="ROLE_1_option"
            value="ROLE_1"
          >
            Group 1
          </option>
          <option
            data-qa="ROLE_2_option"
            value="ROLE_2"
          >
            Group 2
          </option>
          <option
            data-qa="ROLE_3_option"
            value="ROLE_3"
          >
            Group 3
          </option>
        </select>
      </label>
      <div
        className="sc-fzozJi lkMuvo"
      >
        <button
          className="sc-AxgMl dzrPxP"
          data-qa="add-button"
          disabled={false}
          type="submit"
        >
          Add group
        </button>
        <button
          className="sc-AxgMl iErfzZ"
          data-qa="cancel-button"
          disabled={false}
          onClick={[MockFunction]}
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Auth user add groups no groups should render correctly with user and no groups 1`] = `
<div
  className="sc-Axmtr hjHhIB"
>
  <div
    className="sc-AxmLO czuJjP"
  >
    <form
      onSubmit={[Function]}
    >
      <div
        data-qa="no-groups"
      >
        No groups available
      </div>
      <div
        className="sc-fzozJi lkMuvo"
      >
        <button
          className="sc-AxgMl iErfzZ"
          data-qa="cancel-button"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Auth user add groups should render the add group page with a filtered list of groups 1`] = `
<div
  className="sc-Axmtr hjHhIB"
>
  <div
    className="sc-AxmLO czuJjP"
  >
    <form
      onSubmit={[MockFunction]}
    >
      <label
        className="sc-AxjAm bnkTnU"
        htmlFor="group"
      >
        <span
          className="sc-AxirZ lbZqdG"
        >
          Choose new group
        </span>
        <select
          className="sc-AxhUy czrCMV"
          id="group"
          name="group"
          onChange={[MockFunction]}
        >
          <option
            value="--"
          >
            -- Select --
          </option>
          <option
            data-qa="ROLE_1_option"
            value="ROLE_1"
          >
            Group 1
          </option>
          <option
            data-qa="ROLE_2_option"
            value="ROLE_2"
          >
            Group 2
          </option>
          <option
            data-qa="ROLE_3_option"
            value="ROLE_3"
          >
            Group 3
          </option>
        </select>
      </label>
      <div
        className="sc-fzozJi lkMuvo"
      >
        <button
          className="sc-AxgMl dzrPxP"
          data-qa="add-button"
          disabled={false}
          type="submit"
        >
          Add group
        </button>
        <button
          className="sc-AxgMl iErfzZ"
          data-qa="cancel-button"
          disabled={false}
          onClick={[MockFunction]}
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</div>
`;
